### PR TITLE
Producer: Http receiver

### LINKED ISSUE #299: Implement http receiver/producer

### PR DESCRIPTION
**Pull request by vd2org**

### CODE REVIEW

File: core/pyproject.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ cryptography = ">=39.0.1"
 jmespath = "^1.0.0"
 jsonschema = "^4.4.0"
 orjson = "^3.8.7"
-python = "^3.7"
+python = "^3.8"
```
Comment (spicy-sauce) [inline core/pyproject.toml]: we can't do it, as RegisGears runs on Python 3.7.

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        finally:
+            with suppress(Exception):
+                await srv.stop()
+
+    def stop(self):
+        """
+        Cleans the block connections and state
+        """
+        pass
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: If it's not implemented yet, you can remove this completely, as it will inherit it from the parent.

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+class Block(DyProducer, metaclass=ABCMeta):
+    port: int
+
+    def init(self, context: Optional[Context] = None):
+        logger.debug(f"Initializing {self.get_block_name()}")
+        self.port = int(self.properties.get("encoding", 8080))
+
+    async def produce(self) -> AsyncGenerator[List[Message], None]:
+        q = Queue(maxsize=1000)
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: mind to rename variable to `queue`?

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+logger = logging.getLogger("dy")
+
+
+class Block(DyProducer, metaclass=ABCMeta):
+    port: int
+
+    def init(self, context: Optional[Context] = None):
+        logger.debug(f"Initializing {self.get_block_name()}")
+        self.port = int(self.properties.get("encoding", 8080))
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: is it encoding? seems like a mistake...

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    async def produce(self) -> AsyncGenerator[List[Message], None]:
+        q = Queue(maxsize=1000)
+
+        async def handler(request: BaseRequest) -> Response:
+            try:
+                q.put_nowait(orjson.loads(await request.read()))
+                return Response(status=200)
+            except Exception:  # noqa
+                logger.exception("Got exception while parsing request:")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py:35]: can you also log the exception itself?

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                return Response(status=500, text="Server error.")
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: Don't they have a constant for 500 internal error?

File: core/src/datayoga_core/blocks/http/receiver/block.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        logger.debug(f"Initializing {self.get_block_name()}")
+        self.port = int(self.properties.get("encoding", 8080))
+
+    async def produce(self) -> AsyncGenerator[List[Message], None]:
+        q = Queue(maxsize=1000)
+
+        async def handler(request: BaseRequest) -> Response:
+            try:
+                q.put_nowait(orjson.loads(await request.read()))
+                return Response(status=200)
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: Don't they have a constant for 200 ok?

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 4 above)
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py]: better to take the default 8080 from the json schema itself so if we change it to 8081 we won't need to do it twice... This is relevant to other properties and their defaults.

File: dev/compose-mysql.yaml
Comment (spicy-sauce) [inline dev/compose-mysql.yaml:1]: how are those compose files related to this PR?

File: core/src/datayoga_core/blocks/http/receiver/block.py
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.py:1]: can you add tests? maybe by mocking a http server?

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 5 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py:35]: The `exception` method is unwinding current stacktrace implicitly. Don't need to pass exception here.

File: dev/compose-mysql.yaml
Comment (vd2org) [reply dev/compose-mysql.yaml:1]: I improve it as soon as it needs.

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 4 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 3 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 2 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/pyproject.toml (hunk 1 above)
Comment (vd2org) [reply core/pyproject.toml]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 6 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 7 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 4 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: It's hard to do it here, because block don't have access to the schema. Thus, it must be refactored.

File: core/src/datayoga_core/blocks/http/receiver/block.py
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py:1]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 4 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: OK, we probably should take care of it in a different task.

File: core/src/datayoga_core/blocks/http/receiver/block.schema.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    }
+  },
+  "additionalProperties": false,
+  "examples": [
+    {
+      "host": "localhost",
+      "port": 8080
+    }
+  ]
+}
```
Comment (spicy-sauce) [inline core/src/datayoga_core/blocks/http/receiver/block.schema.json:24]: It seems to me that the following properties are missing:
- endpoint/uri: currently it's host:port
- https: boolean
- method: GET/POST, etc.
- Credentials for basic authentication.

File: dev/compose-mysql.yaml
Comment (spicy-sauce) [reply dev/compose-mysql.yaml:1]: OK, but what do they have to do with this PR? Was it taken from the previous unmerged demo PR?

Comment (vd2org) [reply dev/compose-mysql.yaml:1]: Nope. It's just the result of testing something in parallel. Probably WB and DB connectivity.

File: core/src/datayoga_core/blocks/http/receiver/block.py (hunk 4 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.py]: ✅

File: core/src/datayoga_core/blocks/http/receiver/block.schema.json (hunk 8 above)
Comment (vd2org) [reply core/src/datayoga_core/blocks/http/receiver/block.schema.json:24]: I don't think we have to support https nor auth here.
The same applies to the method: it doesn't matter which one is used - producer will accept everything have a body.

Why? It's really hard to implement properly all those thinks. 
I mean, It's not only https, it's also cert based auth, external auth, LDAP and so on.

Instead of reimplementing all of this on our side we just forcing users to use appropriate component(traefik, nginx, etc) on top of runner. 

So, in short, I would like to keep this as simple as possible.

**Review by spicy-sauce — APPROVED** [review]
